### ISSUE TITLE
Language Server

### ISSUE THREAD

**Issue by austince** [processing/processing#5423]
## Description

I'm wondering if there has been any work to create a Language Server for Processing yet. It would be very useful for people who would like to use the language outside of the Processing IDE like [this project for VS Code](https://github.com/TobiahZ/processing-vscode). I'm not sure if I'll have time I'd love to help try to get it started if it hasn't been before!

**Comment by benfry** [comment]
Not that I know of; please let us know if you make some progress. I don't know how many people are using VS Code, but perhaps it'd be helpful? Thanks for checking in.

**Comment by JakubValtar** [comment]
I'm working on a new preprocessor which would allow us to run Java 8. Until then I don't think it's possible to create a language server because our current preprocessor is quite picky about which syntax it will accept (it's something between Java 5 and 8) and it's a bit of a mess.

When we are on Java 8 we will need a new autocompletion engine. Extending Java Language Server and possibly moving more functionality from the PDE there seems to be a good option, but we still have to decide on the architecture (do we use LS, what needs to stay in the PDE vs what can be moved/duplicated to the language server and therefore reusable in other editors).

**Comment by austince** [comment]
Ok, sounds good @JakubValtar. Do you need any help with the preprocessor?

**Comment by neilcsmith-net** [comment]
Just linking two conversations together! :wink: https://forum.processing.org/two/discussion/26740/proposal-development-environment-javascript-prototype

Out of interest, why was the title changed to "for VS Code"?

**Comment by pr1metine** [comment]
Any progress? What about [this](https://github.com/processing-language-server/LS4P)? Can't seem to get it working...

**Comment by jeremydouglass** [comment]
Yes, LS4P was a GSOC 2019 project. Here is the final report:

-  https://syamsundarkirubakaran.github.io/lspfinal
-  https://medium.com/processing-foundation/google-summer-of-code-2019-wrap-up-post-3478323bb0ea

Good point that once that is complete then the issue here could probably be closed.

It seems like you opened an issue there about you having problems building it on a windows hyper-v...?

https://github.com/processing-language-server/LS4P/issues/6

**Comment by pr1metine** [comment]
Yeah, I was unable to run LS4P on a Ubuntu VM and WSL, let alone on native Windows. Did you, jeremydouglass, or anyone else manage to get it up and running? If that's the case I'll know that it's an issue on my part.

**Comment by benfry** [comment]
We'd still like to see LSP support, though I don't have time to build it myself. [Theia](https://github.com/eclipse-theia/theia) also looks promising if we were to replace the PDE entirely.

**Comment by benfry** [comment]
(Moving this to the Processing 4 repo since we wouldn't be coming back to 3.x to do this.)
